distributed/protocol/serialize.py

```diff
@@ -434,33 +430,116 @@ def typename(typ):
     return typ.__module__ + '.' + typ.__name__
 
 
-def _find_lazy_registration(typename):
-    toplevel, _, _ = typename.partition('.')
-    if toplevel in lazy_registrations:
-        lazy_registrations.pop(toplevel)()
-        return True
-    else:
-        return False
-
-
 @partial(normalize_token.register, Serialized)
 def normalize_Serialized(o):
     return [o.header] + o.frames  # for dask.base.tokenize
 
 
 # Teach serialize how to handle bytestrings
+@dask_serialize.register((bytes, bytearray))
 def _serialize_bytes(obj):
     header = {}  # no special metadata
     frames = [obj]
     return header, frames
 
 
+@dask_deserialize.register((bytes, bytearray))
 def _deserialize_bytes(header, frames):
     return frames[0]
 
 
-# NOTE: using the same exact serialization means a bytes object may be
-# deserialized as bytearray or vice-versa...  Not sure this is a problem
-# in practice.
-register_serialization(bytes, _serialize_bytes, _deserialize_bytes)
-register_serialization(bytearray, _serialize_bytes, _deserialize_bytes)
+#########################
+# Descend into __dict__ #
+#########################
+
+
+basic = (numbers.Number, str)
+
+
+def serialize_object_with_dict(est):
+    header = {
+        'serializer': 'dask',
+        'type-serialized': pickle.dumps(type(est)),
+        'simple': {},
+        'complex': {}
+    }
+    frames = []
+
+    if isinstance(est, dict):
+        d = est
+    else:
+        d = est.__dict__
+
+    for k, v in d.items():
+        typ = type(v)
+        if (typ is str or typ is int or typ is float or
+                isinstance(v, dict) and all(typ is str or typ is int or typ is float for x in v.values())
+                                    and all(typ is str for x in v.keys()) or
+                isinstance(v, (list, tuple)) and all(typ is str or typ is int or typ is float for x in v)):
+            header['simple'][k] = v
+        else:
+            if isinstance(v, dict):
+                h, f = serialize_object_with_dict(v)
+            else:
+                h, f = serialize(v)
+            header['complex'][k] = {'header': h,
+                                    'start': len(frames),
+                                    'stop': len(frames) + len(f)}
+            frames += f
+    return header, frames
+
+
+def deserialize_object_with_dict(header, frames):
+    cls = pickle.loads(header['type-serialized'])
+    if issubclass(cls, dict):
+        dd = obj = {}
+    else:
+        obj = object.__new__(cls)
+        dd = obj.__dict__
+    dd.update(header['simple'])
+    for k, d in header['complex'].items():
+        h = d['header']
+        f = frames[d['start']: d['stop']]
+        v = deserialize(h, f)
+        dd[k] = v
+
+    return obj
+
+
+dask_deserialize.register(dict)(deserialize_object_with_dict)
+
+
+def register_attributes(cls):
```

Comment:
This name feels a bit off to me. Not sure if `register_generic` would be better.

Reply:
Done